examples/graphrnn_s.py

```diff
@@ -0,0 +1,362 @@
+
+
+def is_cycle(graph):
+    return len(nx.cycle_basis(graph)) == 1
+
+
+def plot_4_graphs(graphs, title):
+    plt.figure(figsize=(10, 10))
+    plt.title(title)
+    for i, graph in enumerate(graphs):
+        plt.subplot(2, 2, i + 1)
+        nx.draw_spectral(graph, node_size=100)
+    plt.show()
+
+
+class GraphRNN_S(nn.Module):
+    r"""The GraphRNN-S model from the "GraphRNN: Generating Realistic Graphs
+    with Deep Auto-regressive Models" <https://arxiv.org/pdf/1802.08773>`_
+    paper, consisting of a node-level RNN and a edge-level output layer.
+
+    Args:
+        adjacency_size (int): The size of the adjacency vectors.
+        adjacency_embedding_size (int): The size of the embedding of the
+            of the adjacency vectors before feeding it to the RNN cell.
+        hidden_size (int): The size of the hidden vectors of the RNN cell.
+        num_layers (int): The number of stacked RNN layers.
+        output_embedding_size (int): The size of the embedding of the
+            edge level MLP.
+    """
+    def __init__(
+        self,
+        *,
+        adjacency_size: int,
+        adjacency_embedding_size: int,
+        hidden_size: int,
+        num_layers: int,
+        output_embedding_size: int,
+    ):
+        super().__init__()
+        self.adjacency_size = adjacency_size
+        self.num_layers = num_layers
+        self.hidden_size = hidden_size
+        self.hidden = None
+
+        self.embedding = nn.Sequential(
+            nn.Linear(adjacency_size, adjacency_embedding_size),
+            nn.ReLU(),
+        )
+        self.rnn = nn.RNN(
+            input_size=adjacency_embedding_size,
+            hidden_size=hidden_size,
+            num_layers=num_layers,
+            batch_first=True,
+        )
+        self.adjacency_mlp = nn.Sequential(
+            nn.Linear(hidden_size, output_embedding_size),
+            nn.ReLU(),
+            nn.Linear(output_embedding_size, adjacency_size),
+            nn.Sigmoid(),
+        )
+
+    def mask_out_bits_after_length(self, sequences, lengths):
+        sequences = rnnutils.pack_padded_sequence(sequences, lengths,
+                                                  batch_first=True,
+                                                  enforce_sorted=False)
+        sequences = rnnutils.pad_packed_sequence(sequences,
+                                                 batch_first=True)[0]
+        return sequences
+
+    def forward(self, input_sequences, input_length, sampling=False):
+        r"""Forward pass of the GraphRNN-S model.
+
+        Args:
+            input_sequences (Tensor): For each graph in the batch, the sequence
+                of adjacency vectors (including the first SOS token). Shape:
+                :obj:`(batch_size, max_num_nodes, adjacency_size)`
+            input_length (Tensor): For each graph in the batch, the length of
+                the sequence of adjacency vectors. Shape: :obj:`(batch_size,)`
```

Comment:
sampling missing from docstring btw (I noticed because I don't understand what it does 😄

Reply:
Renamed to `hidden`